File: 0063-unique-paths-ii/0063-unique-paths-ii.java
class Solution {
    int help(int m , int n , int[][] dp){
       if(m< 0 || n < 0 || dp[m][n] == 0) return 0;
       if(m == 0 && n == 0) return 1;

       if(dp[m][n] != -1) return dp[m][n];

       return dp[m][n] = help(m-1,n,dp) + help(m,n-1,dp);
    }
    public int uniquePathsWithObstacles(int[][] gird) {
        int m =gird.length;
        int n = gird[0].length;
        int[][] dp = new int[m][n];

        for(int[] a : dp){
          Arrays.fill(a,-1);
        }
        for(int i =0;i<m;i++){
            for(int j =0;j<n;j++){
              if(gird[i][j] == 1){
                dp[i][j] = 0;
              }
            }
        }
        return help(m-1,n-1,dp);

        
    }
}

File: 0063-unique-paths-ii/0063-unique-paths-ii.cpp
class Solution {
public:
int helper(int i, int j, vector<vector<int>>& dp,vector<vector<int>>& arr){

    if(i == 0 && j == 0) return 1;
    if(i<0 || j<0 || arr[i][j] == 1) return 0;
      if(dp[i][j] != -1) return dp[i][j];
    int up = helper(i-1,j,dp,arr);
    int back = helper(i,j-1,dp,arr);

    return dp[i][j] = up + back;
}
    int uniquePathsWithObstacles(vector<vector<int>>& arr) {

        if(arr[0][0] == 1) return 0;
        vector<vector<int>>dp(arr.size(),vector<int>(arr[0].size(),-1));
        return helper(arr.size()-1,arr[0].size()-1,dp,arr);
    }
};